num/src/num.rs

```diff
@@ -0,0 +1,306 @@
+    let bits = val.to_bits();
+    let mut shifted_bits = bits >> 8;
+    let lost_bits = bits & 0xFF;
+    // round ties to even
+    shifted_bits += u32::from(!val.is_nan()) & ((lost_bits - ((lost_bits >> 7) & !shifted_bits)) >> 7);
+    // ensure NaNs don't become infinities
+    shifted_bits |= u32::from(val.is_nan());
+    unsafe { Self::new(NumTag::F24, shifted_bits) }
+  }
+
+  #[inline(always)]
+  pub const fn new_sym(val: NumTag) -> Self {
+    unsafe { Self::new(NumTag::Sym, val as u32) }
+  }
+
+  #[inline(always)]
+  pub fn raw(self) -> u32 {
+    self.raw
+  }
+
+  #[inline(always)]
+  pub fn tag(self) -> NumTag {
+    unsafe { NumTag::from_unchecked((self.raw >> 3 & 0x1f) as u8) }
+  }
+
+  #[inline(always)]
+  pub fn payload(self) -> u32 {
+    self.raw >> 8
+  }
+
+  #[inline(always)]
+  pub fn get_u24(self) -> u32 {
+    self.payload()
+  }
+
+  #[inline(always)]
+  pub fn get_i24(self) -> i32 {
+    (self.raw as i32) >> 8
+  }
+
+  #[inline(always)]
+  pub fn get_f24(self) -> f32 {
+    f32::from_bits(self.raw & !0xff)
+  }
+
+  #[inline(always)]
+  pub unsafe fn get_sym(self) -> NumTag {
+    unsafe { NumTag::from_unchecked(self.get_u24() as u8) }
+  }
+
+  #[inline]
+  pub fn operate_unary(op: NumTag, num: Self) -> Self {
+    const U24_MAX: u32 = (1 << 24) - 1;
+    const U24_MIN: u32 = 0;
+    const I24_MAX: i32 = (1 << 23) - 1;
+    const I24_MIN: i32 = (-1) << 23;
+
+    match op {
+      NumTag::Sym => num,
+      NumTag::U24 => match num.tag() {
+        NumTag::U24 => num,
+        NumTag::I24 => Num::new_u24(num.get_u24()),
+        NumTag::F24 => Num::new_u24((num.get_f24() as u32).clamp(U24_MIN, U24_MAX)),
+        _ => Self::INVALID,
+      },
+      NumTag::I24 => match num.tag() {
+        NumTag::U24 => Num::new_i24(num.get_i24()),
+        NumTag::I24 => num,
+        NumTag::F24 => Num::new_i24((num.get_f24() as i32).clamp(I24_MIN, I24_MAX)),
+        _ => Self::INVALID,
+      },
+      NumTag::F24 => match num.tag() {
+        NumTag::U24 => Num::new_f24(num.get_u24() as f32),
+        NumTag::I24 => Num::new_f24(num.get_i24() as f32),
+        NumTag::F24 => num,
+        _ => Self::INVALID,
+      },
+      _ => unsafe { Self::new(op, num.payload()) },
+    }
+  }
+
+  #[inline]
+  pub fn operate_sym(a: Self, b: Self) -> Self {
+    let at = a.tag();
+    let bt = b.tag();
+    let (op, ty, a, b) = match ((at, at.is_op(), a), (bt, bt.is_op(), b)) {
+      ((NumTag::Sym, ..), (NumTag::Sym, ..)) => return Self::INVALID,
+      sym!((NumTag::Sym, _, a), (.., b)) => return Self::operate_unary(unsafe { a.get_sym() }, b),
+      ((_, false, _), (_, false, _)) | ((_, true, _), (_, true, _)) => return Self::INVALID,
+      sym!((op, true, a), (ty, false, b)) => (op, ty, a, b),
+    };
+    match ty {
+      NumTag::U24 => {
+        let a = a.get_u24();
+        let b = b.get_u24();
+        match op {
+          NumTag::U24 | NumTag::I24 | NumTag::F24 | NumTag::Sym => unsafe { unreachable_unchecked() },
```

Comment:
We added the casting operators so i think this is stale

Reply:
The casting operators are handled by `operate_unary`, which is called on line 169.